docs/proposals/scheduling/20220701-pod-migration-job.md

```diff
@@ -0,0 +1,354 @@
+- Check the number of Pods being migrated in the Namespace where each target Pod is located. If it exceeds the maximum migration amount for a single Namespace, exclude it
+
+The detailed algorithm of Workload Max Migrating/Unavailable Replicas:
+
+```go
+func GetMaxMigrating(replicas int, intOrPercent *intstr.IntOrString) (int, error) {
+	return GetMaxUnavailable(replicas, intOrPercent)
+}
+
+func GetMaxUnavailable(replicas int, intOrPercent *intstr.IntOrString) (int, error) {
+	if intOrPercent == nil {
+		if replicas > 10 {
+			s := intstr.FromString("10%")
+			intOrPercent = &s
+		} else if replicas >= 4 && replicas <= 10 {
+			s := intstr.FromInt(2)
+			intOrPercent = &s
+		} else {
+			s := intstr.FromInt(1)
+			intOrPercent = &s
+		}
+	}
+	return intstr.GetValueFromIntOrPercent(intOrPercent, replicas, true)
+}
+``` 
+
+##### Sort PodMigrationJob
+
+- Pods with higher QoS requirements are given priority, LSE > LSR > LS > BE
+- Pods with higher priority will be processed first
+- The higher migration priority will be processed first
+- If the Pod has already initiated a migration job in the past and it fails, sort by the number of times. The lower the number of times, the priority will be given to processing
+- If the workload where the Pod is located has been descheduled for a certain number of times in the past, it is sorted according to the number of times. The lower the number of times, the priority will be processed.
+- Sort by the number of replicas being migrated by the workload. The lower the number of replicas being migrated, the priority will be given to processing.
+
+##### Execute PodMigrationJob
+
+- Update PodMigrationJobStatus.Phase to Running to trigger the PodMigrationJob controller reconcile these jobs
+- PodMigrationJob controller reconciles process:
+  - If the mode of PodMigrationJob is `DeleteDirectly`, just delete the Pod through the delete method that configured in PodMigrationJob controller. And update the phase of PodMigrationJob to Success.
+  - If not specified ReservationOptions.ReservationRef, create the Reservation instance by the reservation template or Pod  spec to reserve resources. And updates the created Reservation instance to the ReservationOptions.ReservationRef.
+  - Check the status of Reservation to determine whether reserve resource successfully.
+  - If failed to reserve, abort the PodMigrationJob and update the phase of PodMigrationJob to Fail
+  - If successfully reserve, delete the Pod through the delete method that configured in PodMigrationJob controller.
+  - Check the Reservation status to determine whether the Reservation consumed.
+  - If Reservation consumed, tracks the status of Reservation and update the status to PodMigrationJob
+  - Update phase of PodMigrationJob to Success.
+
+##### Migration Stability mechanism
+
+- Support for disabling this capability by configuration 
+- Supports a simple central flow control mechanism to limit the number of migrations over a period of time.
```

Comment:
Add some description, how the user configures this security mechanism (boot parameters, configmap, or something else?)

Reply:
Added relevant details.

Reply:
Added relevant details.